fix(build): Migrate from tsc to tsup for proper ESM module resolution

## Problem

When consuming this package in Vitest with `happy-dom` or `jsdom` environment, the following error occurs:

```
Error: Cannot find module '/path/to/node_modules/@laststance/redux-storage-middleware/dist/storageMiddleware' 
imported from '/path/to/node_modules/@laststance/redux-storage-middleware/dist/index.js'
```

### Root Cause

The current build configuration uses TypeScript compiler (`tsc`) directly with `moduleResolution: "bundler"`:

**tsconfig.json:**
```json
{
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "bundler",
    ...
  }
}
```

**package.json:**
```json
{
  "type": "module",
  "scripts": {
    "build": "tsc"
  }
}
```

This combination produces output without `.js` extensions:

**dist/index.js (current):**
```js
export { createStorageMiddleware, ... } from './storageMiddleware';  // ❌ No .js extension
export { createSafeLocalStorage, ... } from './storage';             // ❌ No .js extension
```

Node.js ESM strict resolution requires explicit `.js` extensions:
```js
export { createStorageMiddleware, ... } from './storageMiddleware.js';  // ✅ Required
```

### Why This Matters

| Environment | Works? | Reason |
|------------|--------|--------|
| Bundlers (Vite, webpack, esbuild) | ✅ | Bundlers resolve extensionless imports |
| Node.js ESM direct | ❌ | Strict ESM requires `.js` extensions |
| Vitest (happy-dom/jsdom) | ❌ | Uses Node.js module resolution for `node_modules` |
| Vitest (browser mode) | ✅ | Uses Vite bundler |

---

## Solution: Migrate to tsup

[tsup](https://github.com/egoist/tsup) is a zero-config TypeScript bundler powered by esbuild that:
- Automatically handles ESM module resolution
- Supports dual ESM/CJS output
- Generates TypeScript declarations
- Much faster than `tsc`

---

## Implementation Steps

### Step 1: Install tsup

```bash
pnpm add -D tsup
```

### Step 2: Create tsup.config.ts

```ts
import { defineConfig } from 'tsup'

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['esm', 'cjs'],      // Dual format for maximum compatibility
  dts: true,                    // Generate .d.ts files
  splitting: false,             // Don't split chunks (single entry)
  sourcemap: true,              // Keep sourcemaps
  clean: true,                  // Clean dist/ before build
  outExtension({ format }) {
    return {
      js: format === 'esm' ? '.mjs' : '.cjs',
    }
  },
})
```

### Step 3: Update package.json

```diff
{
  "name": "@laststance/redux-storage-middleware",
  "version": "0.2.1",
  "type": "module",
- "main": "dist/index.js",
- "module": "dist/index.js",
- "types": "dist/index.d.ts",
+ "main": "dist/index.cjs",
+ "module": "dist/index.mjs",
+ "types": "dist/index.d.ts",
  "exports": {
    ".": {
-     "types": "./dist/index.d.ts",
-     "import": "./dist/index.js",
-     "require": "./dist/index.js"
+     "types": "./dist/index.d.ts",
+     "import": "./dist/index.mjs",
+     "require": "./dist/index.cjs"
    }
  },
  "scripts": {
-   "build": "tsc",
+   "build": "tsup",
  }
}
```

### Step 4: Update tsconfig.json (for type checking only)

```diff
{
  "compilerOptions": {
    "target": "ES2020",
    "module": "ESNext",
    "lib": ["ES2020", "DOM"],
-   "moduleResolution": "bundler",
+   "moduleResolution": "NodeNext",
    "declaration": true,
    "declarationMap": true,
    "sourceMap": true,
-   "outDir": "./dist",
+   "noEmit": true,
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "isolatedModules": true
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules", "dist", "examples", "**/*.test.ts"]
}
```

**Note:** `noEmit: true` because tsup handles the actual compilation, while `tsc --noEmit` is used for type checking only.

### Step 5: Update scripts

```json
{
  "scripts": {
    "clean": "rm -rf dist/",
    "prebuild": "pnpm clean",
    "build": "tsup",
    "typecheck": "tsc --noEmit"
  }
}
```

---

## Expected Output After Migration

**dist/index.mjs (ESM):**
```js
export { createStorageMiddleware, ... } from './storageMiddleware.mjs';  // ✅ Proper extension
```

**dist/index.cjs (CJS):**
```js
const { createStorageMiddleware, ... } = require('./storageMiddleware.cjs');  // ✅ CJS format
```

---

## Verification Checklist

After migration, verify:

- [ ] `pnpm build` succeeds
- [ ] `pnpm typecheck` passes
- [ ] `pnpm test` passes (all 35 tests)
- [ ] `dist/` contains `.mjs`, `.cjs`, and `.d.ts` files
- [ ] Consuming projects work without `test.server.deps.inline` workaround
- [ ] Run `attw` (Are the Types Wrong?) to validate exports

### Verification with attw

```bash
npx @arethetypeswrong/cli --pack .
```

Expected output should show ✅ for all resolution modes.

---

## Benefits

| Aspect | Before (tsc) | After (tsup) |
|--------|--------------|--------------|
| Build time | ~2s | ~0.3s |
| ESM compatibility | ❌ Broken | ✅ Works |
| CJS compatibility | ❌ Missing | ✅ Added |
| Bundler integration | ✅ | ✅ |
| Node.js direct | ❌ | ✅ |
| Vitest (Node env) | ❌ | ✅ |

---

## References

- [tsup documentation](https://tsup.egoist.dev/)
- [Dual Publishing ESM and CJS Modules with tsup](https://johnnyreilly.com/dual-publishing-esm-cjs-modules-with-tsup-and-are-the-types-wrong)
- [TypeScript in 2025: ESM and CJS npm publishing](https://lirantal.com/blog/typescript-in-2025-with-esm-and-cjs-npm-publishing)
- [Are the Types Wrong?](https://github.com/arethetypeswrong/arethetypeswrong.github.io)

---

## Related

This issue was discovered while fixing Vitest test failures in a consuming project (gitbox). The current workaround requires consumers to add:

```ts
// vitest.config.ts
test: {
  server: {
    deps: {
      inline: ['@laststance/redux-storage-middleware'],
    },
  },
}
```

This migration will eliminate the need for such workarounds.

--- a/eslint.config.js
+++ b/eslint.config.js
@@ -13,6 +13,7 @@ export default defineConfig([
       'benchmarks/**',
       'eslint.config.js',
       'vitest.config.ts',
+      'tsup.config.ts',
     ],
   },
   ...tsPrefixer,

--- a/package.json
+++ b/package.json
@@ -3,14 +3,14 @@
   "version": "0.2.0",
   "type": "module",
   "description": "SSR-safe Redux middleware for LocalStorage persistence with hydration control",
-  "main": "dist/index.js",
-  "module": "dist/index.js",
-  "types": "dist/index.d.ts",
+  "main": "./dist/index.cjs",
+  "module": "./dist/index.mjs",
+  "types": "./dist/index.d.ts",
   "exports": {
     ".": {
       "types": "./dist/index.d.ts",
-      "import": "./dist/index.js",
-      "require": "./dist/index.js"
+      "import": "./dist/index.mjs",
+      "require": "./dist/index.cjs"
     }
   },
   "files": [
@@ -23,9 +23,7 @@
     "*": "prettier --ignore-unknown --write"
   },
   "scripts": {
-    "clean": "rm -rf dist/",
-    "prebuild": "pnpm clean",
-    "build": "tsc",
+    "build": "tsup",
     "test": "vitest",
     "test:run": "vitest run",
     "test:coverage": "vitest run --coverage",
@@ -72,6 +70,7 @@
     "lint-staged": "^16.2.7",
     "playwright": "^1.57.0",
     "prettier": "^3.7.4",
+    "tsup": "^8.5.1",
     "typescript": "^5.3.0",
     "vitest": "^4.0.16"
   }

--- a/src/index.ts
+++ b/src/index.ts
@@ -45,7 +45,7 @@ export {
   ACTION_HYDRATE_START,
   ACTION_HYDRATE_COMPLETE,
   ACTION_HYDRATE_ERROR,
-} from './storageMiddleware'
+} from './storageMiddleware.js'
 
 // =============================================================================
 // Storage
@@ -60,7 +60,7 @@ export {
   isValidStorage,
   getStorageSize,
   getRemainingStorageQuota,
-} from './storage'
+} from './storage.js'
 
 // =============================================================================
 // Serializers
@@ -83,7 +83,7 @@ export {
   getCompressionRatio,
   type CompressionFormat,
   type CompressedSerializerOptions,
-} from './serializers'
+} from './serializers/index.js'
 
 // =============================================================================
 // Utilities
@@ -98,7 +98,7 @@ export {
   debounceLeading,
   throttle,
   scheduleIdleCallback,
-} from './utils'
+} from './utils/index.js'
 
 // =============================================================================
 // Types
@@ -129,4 +129,4 @@ export type {
   // Utility Types
   NestedKeyOf,
   PathValue,
-} from './types'
+} from './types.js'

--- a/src/serializers/compressed.ts
+++ b/src/serializers/compressed.ts
@@ -8,8 +8,8 @@
  * pnpm add lz-string
  */
 
-import type { Serializer } from '../types'
-import { createModuleLoader } from '../utils/moduleLoader'
+import type { Serializer } from '../types.js'
+import { createModuleLoader } from '../utils/moduleLoader.js'
 
 /**
  * LZString interface (for dynamic import)

--- a/src/serializers/index.ts
+++ b/src/serializers/index.ts
@@ -12,13 +12,13 @@ export {
   dateReviver,
   collectionReplacer,
   collectionReviver,
-} from './json'
+} from './json.js'
 
 export {
   createSuperJsonSerializer,
   initSuperJsonSerializer,
   isSuperJsonLoaded,
-} from './superjson'
+} from './superjson.js'
 
 export {
   createCompressedSerializer,
@@ -27,4 +27,4 @@ export {
   getCompressionRatio,
   type CompressionFormat,
   type CompressedSerializerOptions,
-} from './compressed'
+} from './compressed.js'

--- a/src/serializers/json.ts
+++ b/src/serializers/json.ts
@@ -4,7 +4,7 @@
  * Serializer using standard JSON.stringify/parse
  */
 
-import type { Serializer, JsonSerializerOptions } from '../types'
+import type { Serializer, JsonSerializerOptions } from '../types.js'
 
 /**
  * Set of dangerous keys to prevent prototype pollution attacks

--- a/src/serializers/superjson.ts
+++ b/src/serializers/superjson.ts
@@ -8,8 +8,8 @@
  * pnpm add superjson
  */
 
-import type { Serializer } from '../types'
-import { createModuleLoader } from '../utils/moduleLoader'
+import type { Serializer } from '../types.js'
+import { createModuleLoader } from '../utils/moduleLoader.js'
 
 /**
  * SuperJSON interface (for dynamic import)

--- a/src/storage.ts
+++ b/src/storage.ts
@@ -5,8 +5,11 @@
  * Prevents crashes in SSR environments and supports custom storage backends
  */
 
-import type { StateStorage, SyncStorage, AsyncStorage } from './types'
-import { isStorageAvailable, isSessionStorageAvailable } from './utils/isServer'
+import type { StateStorage, SyncStorage, AsyncStorage } from './types.js'
+import {
+  isStorageAvailable,
+  isSessionStorageAvailable,
+} from './utils/isServer.js'
 
 // =============================================================================
 // Constants

--- a/src/storageMiddleware.ts
+++ b/src/storageMiddleware.ts
@@ -13,18 +13,18 @@ import type {
   Reducer,
 } from '@reduxjs/toolkit'
 
-import { defaultJsonSerializer } from './serializers/json'
-import { createSafeLocalStorage } from './storage'
+import { defaultJsonSerializer } from './serializers/json.js'
+import { createSafeLocalStorage } from './storage.js'
 import type {
   StorageMiddlewareConfig,
   StorageMiddlewareResult,
   HydrationApi,
   HydrationState,
   PersistedState,
-} from './types'
-import { debounce } from './utils/debounce'
-import { isServer, isStorageAvailable } from './utils/isServer'
-import { throttle, scheduleIdleCallback } from './utils/throttle'
+} from './types.js'
+import { debounce } from './utils/debounce.js'
+import { isServer, isStorageAvailable } from './utils/isServer.js'
+import { throttle, scheduleIdleCallback } from './utils/throttle.js'
 
 // =============================================================================
 // Constants

--- a/src/utils/index.ts
+++ b/src/utils/index.ts
@@ -9,8 +9,8 @@ export {
   isBrowser,
   isStorageAvailable,
   isSessionStorageAvailable,
-} from './isServer'
+} from './isServer.js'
 
-export { debounce, debounceLeading } from './debounce'
+export { debounce, debounceLeading } from './debounce.js'
 
-export { throttle, scheduleIdleCallback } from './throttle'
+export { throttle, scheduleIdleCallback } from './throttle.js'

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -1,13 +1,10 @@
 {
   "compilerOptions": {
     "target": "ES2020",
-    "module": "ESNext",
+    "module": "NodeNext",
     "lib": ["ES2020", "DOM"],
-    "moduleResolution": "bundler",
-    "declaration": true,
-    "declarationMap": true,
-    "sourceMap": true,
-    "outDir": "./dist",
+    "moduleResolution": "NodeNext",
+    "noEmit": true,
     "strict": true,
     "esModuleInterop": true,
     "skipLibCheck": true,

--- a/tsup.config.ts
+++ b/tsup.config.ts
@@ -1,0 +1,16 @@
+import { defineConfig } from 'tsup'
+
+export default defineConfig({
+  entry: ['src/index.ts'],
+  format: ['esm', 'cjs'],
+  dts: true,
+  splitting: false,
+  sourcemap: true,
+  clean: true,
+  external: ['@reduxjs/toolkit', 'react-redux', 'lz-string', 'superjson'],
+  outExtension({ format }) {
+    return {
+      js: format === 'esm' ? '.mjs' : '.cjs',
+    }
+  },
+})

